src/main/java/ca/uhn/fhir/jpa/starter/CustomConsentService.java

```diff
@@ -0,0 +1,129 @@
+package ca.uhn.fhir.jpa.starter;
+
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.IdType;
+import org.hl7.fhir.r4.model.Reference;
+import org.hl7.fhir.r4.model.Resource;
+import org.springframework.stereotype.Service;
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import ca.uhn.fhir.jpa.api.dao.DaoRegistry;
+import ca.uhn.fhir.rest.api.server.RequestDetails;
+import ca.uhn.fhir.rest.server.interceptor.consent.ConsentOutcome;
+import ca.uhn.fhir.rest.server.interceptor.consent.IConsentContextServices;
+import ca.uhn.fhir.rest.server.interceptor.consent.IConsentService;
+
+@Service
+public class CustomConsentService implements IConsentService {
+
+  private static final String OAUTH_CLAIM_NAME = "patient";
+
+  private static final org.slf4j.Logger log =
+      org.slf4j.LoggerFactory.getLogger(CustomConsentService.class);
+
+  private OAuth2Helper oAuth2Helper = new OAuth2Helper();
+  DaoRegistry daoRegistry;
+
+  public CustomConsentService() {}
+
+  public CustomConsentService(DaoRegistry daoRegistry) {
+    this.daoRegistry = daoRegistry;
+  }
+
+  @Override
+  public ConsentOutcome startOperation(RequestDetails theRequestDetails,
+      IConsentContextServices theContextServices) {
+    if (oAuth2Helper.checkInPatientCompartment(theRequestDetails.getResourceName())) {
+      return ConsentOutcome.AUTHORIZED;
+    }
+    if (theRequestDetails.getHeader("Authorization") != null) {
+      String patientId = getPatientFromToken(theRequestDetails);
+      String resourceName = null;
+      boolean proceed = false;
+      switch (theRequestDetails.getRequestType().toString()) {
+        case "POST":
+          resourceName = theRequestDetails.getResourceName();
+          proceed = validateResource(resourceName, patientId, theRequestDetails.getResource());
+          break;
+        case "PUT":
+          resourceName = theRequestDetails.getResourceName();
+          IBaseResource putResource = getResourceFromDB(theRequestDetails.getRequestPath());
+          if (validateResource(resourceName, patientId, putResource)) {
+            proceed = validateResource(resourceName, patientId, theRequestDetails.getResource());
+          }
+          break;
+        case "PATCH":
+          resourceName = theRequestDetails.getResourceName();
+          IBaseResource patchResource = getResourceFromDB(theRequestDetails.getRequestPath());
+          proceed = validateResource(resourceName, patientId, patchResource);
+          break;
+        case "GET":
+          resourceName = theRequestDetails.getResourceName();
+          if (theRequestDetails.getRequestPath().split("/").length > 1) {
```

Comment:
What is this line doing? The code doesn't make it clear to me.

Reply:
It checks the request path having resource and its Id.